p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,275 @@
+
+const (
+	maxAcceptQueueLen int = 10
+)
+
+type connection struct {
+	pc        *webrtc.PeerConnection
+	transport *WebRTCTransport
+	scope     network.ConnManagementScope
+
+	localPeer      peer.ID
+	privKey        ic.PrivKey
+	localMultiaddr ma.Multiaddr
+
+	remotePeer      peer.ID
+	remoteKey       ic.PubKey
+	remoteMultiaddr ma.Multiaddr
+
+	m       sync.Mutex
+	streams map[uint16]*dataChannel
+
+	acceptQueue chan network.MuxedStream
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func newConnection(
+	pc *webrtc.PeerConnection,
+	transport *WebRTCTransport,
+	scope network.ConnManagementScope,
+
+	localPeer peer.ID,
+	privKey ic.PrivKey,
+	localMultiaddr ma.Multiaddr,
+
+	remotePeer peer.ID,
+	remoteKey ic.PubKey,
+	remoteMultiaddr ma.Multiaddr,
+) *connection {
+	acceptQueue := make(chan network.MuxedStream, maxAcceptQueueLen)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	conn := &connection{
+		pc:        pc,
+		transport: transport,
+		scope:     scope,
+
+		localPeer:      localPeer,
+		privKey:        privKey,
+		localMultiaddr: localMultiaddr,
+
+		remotePeer:      remotePeer,
+		remoteKey:       remoteKey,
+		remoteMultiaddr: remoteMultiaddr,
+		ctx:             ctx,
+		cancel:          cancel,
+		streams:         make(map[uint16]*dataChannel),
+
+		acceptQueue: acceptQueue,
+	}
+
+	pc.OnDataChannel(func(dc *webrtc.DataChannel) {
+		id := *dc.ID()
+		var stream *dataChannel
+		dc.OnOpen(func() {
+			// datachannel cannot be detached before opening
+			rwc, err := dc.Detach()
+			if err != nil {
+				log.Errorf("[%s] could not detach channel: %s", localPeer, dc.Label())
+				return
+			}
+			stream = newDataChannel(conn, dc, rwc, pc, nil, nil)
+			select {
+			case acceptQueue <- stream:
+				conn.addStream(id, stream)
+			default:
+				log.Warn("cannot buffer incoming stream, will reset")
+				stream.Reset()
+			}
+
+		})
+
+	})
+
+	pc.OnConnectionStateChange(func(state webrtc.PeerConnectionState) {
+		switch state {
+		case webrtc.PeerConnectionStateDisconnected:
+			log.Warnf("[%s] peer connection disconnected", localPeer)
+		case webrtc.PeerConnectionStateFailed:
+			log.Warn("peer connection reset")
+			conn.resetStreams()
+			fallthrough
+		case webrtc.PeerConnectionStateClosed:
+			log.Warnf("[%s] peer connection closed", localPeer)
+			conn.Close()
+		}
+	})
+
+	return conn
+}
+
+func (c *connection) resetStreams() {
+	if c.IsClosed() {
+		return
+	}
+	for _, stream := range c.streams {
+		stream.Reset()
+	}
+}
+
+// ConnState implements transport.CapableConn
+func (c *connection) ConnState() network.ConnectionState {
+	return network.ConnectionState{
+		Transport: "webrtc",
+	}
+}
+
+// Implement network.MuxedConn
+
+func (c *connection) Close() error {
+	if c.IsClosed() {
+		return nil
+	}
+
+	c.scope.Done()
+	c.cancel()
+	return c.pc.Close()
+}
+
+func (c *connection) IsClosed() bool {
+	select {
+	case <-c.ctx.Done():
+		return true
+	default:
+	}
+	return false
```

Comment:
might as well have it as the body for the default case!?,  no need to have it out o the select sttement